src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -69,6 +67,96 @@ def __init__(
         self.upgrade_strategy = upgrade_strategy
         self._result: Optional[Result] = None
 
+    def _get_ireq(
+        self,
+        candidate: Candidate,
+        direct_url_requested: bool,
+    ) -> Optional[InstallRequirement]:
+        """Get the InstallRequirement to install for a candidate.
+
+        Returning None means the candidate is already satisfied by the current
+        environment state and does not need to be handled.
+        """
+        ireq = candidate.get_install_requirement()
+
+        # No ireq to install (e.g. extra-ed candidate). Skip.
+        if ireq is None:
+            return None
+
+        # The currently installed distribution of the same identifier.
+        installed_dist = self.factory.get_dist_to_uninstall(candidate)
+
+        if installed_dist is None:  # Not installed. Install incoming candidate.
+            return ireq
+
+        # If we return this ireq, it should trigger uninstallation of the
+        # existing distribution for reinstallation.
+        ireq.should_reinstall = True
+
+        # Reinstall if --force-reinstall is set.
+        if self.factory.force_reinstall:
+            return ireq
+
+        # The artifact represented by the incoming candidate.
+        cand_link = candidate.source_link
+
+        # The candidate does not point to an artifact. This means the resolver
+        # has already decided the installed distribution is good enough.
+        if cand_link is None:
+            return None
+
+        # The incoming candidate was produced only from version requirements.
+        # Reinstall if the installed distribution's version does not match.
+        if not direct_url_requested:
+            if installed_dist.version == candidate.version:
+                return None
+            return ireq
+
+        # At this point, the incoming candidate was produced from a direct URL.
+        # Determine whether to upgrade based on flags and whether the installed
+        # distribution was done via a direct URL.
+
+        # Always reinstall a direct candidate if it's on the local file system.
+        if cand_link.is_file:
+            return ireq
+
+        # Reinstall if --upgrade is specified.
+        if self.upgrade_strategy != "to-satisfy-only":
+            return ireq
+
+        # The PEP 610 direct URL representation of the installed distribution.
+        dist_direct_url = installed_dist.direct_url
+
+        # The incoming candidate was produced from a direct URL, but the
+        # currently installed distribution was not. Always reinstall to be sure.
+        if dist_direct_url is None:
+            return ireq
+
+        # Editable candidate always triggers reinstallation.
+        if candidate.is_editable:
+            return ireq
+
+        # The currently installed distribution is editable, but the incoming
+        # candidate is not. Uninstall the editable one to match.
+        if installed_dist.editable:
+            return ireq
```

Comment:
I continue pondering this topic and I tried installing an editable then the same from index.
I was expecting to hit this line, but it did not, because `candidate.get_install_requirement()` returns None which makes `_get_ireq` exit immediately. Instead we get "Requirement already satisfied".

Reply:
By "from index", you mean a `pip install -e .` installing a package named `coffee` and a `pip install coffee` returning `Requirement already statisfied`?

If yes, that's the behaviour I'd expect.

Reply:
> By "from index", you mean a pip install -e . installing a package named coffee and a pip install coffee returning Requirement already statisfied and no reinstall?

That is what I meant, yes.

> If yes, that's the behaviour I'd expect.

Fair. But that is not the behaviour described in the comment at line 139.

Reply:
It is, because on line 113 we already established the incoming candidate is URL-specified. The above scenario would have fell into the block around line 108.

This block can only be triggered by commands like this:

```
$ pip install -e .
$ pip install .  # Or any kind of non-editable URL specification.
```